devices/src/virtio/queue.rs

```diff
@@ -340,6 +340,44 @@ impl Queue {
         }
     }
 
+    /// Returns the number of yet-to-be-popped descriptor chains in the avail ring.
+    pub fn len(&self, mem: &GuestMemory) -> u16 {
+        (self.avail_idx(mem) - self.next_avail).0
+    }
+
+    /// Checks if the driver has made any descriptor chains available in the avail ring.
+    pub fn is_empty(&self, mem: &GuestMemory) -> bool {
+        self.len(mem) == 0
+    }
+
+    /// Pop the first available descriptor chain from the avail ring.
+    pub fn pop<'a, 'b>(&'a mut self, mem: &'b GuestMemory) -> Option<DescriptorChain<'b>> {
+        if self.len(mem) == 0 {
+            return None;
+        }
+        let index_offset = 4 + 2 * (self.next_avail.0 % self.actual_size());
```

Comment:
+1, I don't understand the math here.

Reply:
`index_offset` is the offset into `struct virtq_avail` (in bytes), of the next available descriptor index. I.e. an index into the descriptor table. So, reading the actual descriptor involves two memory hops / dereferences: one for reading the index (located at `virtq_avail->ring[self.next_avail]`), and another for reading the descriptor with that index (`descriptor_table[index]`).

`struct virtq_avail` is defined by the virtio standard as:

```C
struct virtq_avail { 
#define VIRTQ_AVAIL_F_NO_INTERRUPT      1 
        le16 flags; 
        le16 idx; 
        le16 ring[ /* Queue Size */ ]; 
        le16 used_event; /* Only if VIRTIO_F_EVENT_IDX */ 
}; 
```

The index we're looking for is the `self.next_avail`th element of `virtq_avail->ring`, whose offset into `virtq_avail` is `4 + 2 * index` bytes (`flags` and `idx` are 4 bytes in total, and each `ring` entry is 2 bytes).

I'll get this into a code comment as well, since it does look pretty cryptic. Sorry :(